cmd/weave-npc/main.go

```diff
@@ -0,0 +1,160 @@
+		return err
+	}
+
+	if err := ensureFlushedChain(ipt, weavenpc.DefaultChain); err != nil {
+		return err
+	}
+
+	if err := ensureFlushedChain(ipt, weavenpc.MainChain); err != nil {
+		return err
+	}
+
+	// Configure main chain static rules
+	if _, err := ipt.EnsureRule(iptables.Append, iptables.TableFilter, weavenpc.MainChain,
+		"-m", "state", "--state", "RELATED,ESTABLISHED", "-j", "ACCEPT"); err != nil {
+		return err
+	}
+
+	if _, err := ipt.EnsureRule(iptables.Append, iptables.TableFilter, weavenpc.MainChain,
+		"-m", "state", "--state", "NEW", "-j", "WEAVE-NPC-DEFAULT"); err != nil {
+		return err
+	}
+
+	if _, err := ipt.EnsureRule(iptables.Append, iptables.TableFilter, weavenpc.MainChain,
+		"-m", "state", "--state", "NEW", "-j", "WEAVE-NPC-INGRESS"); err != nil {
+		return err
+	}
+
+	if _, err := ipt.EnsureRule(iptables.Append, iptables.TableFilter, weavenpc.MainChain,
+		"-j", "DROP"); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func resetIPSets(ips ipset.Interface) error {
+	// TODO should restrict ipset operations to the `weave-` prefix:
+
+	if err := ips.FlushAll(); err != nil {
+		return err
+	}
+
+	if err := ips.DestroyAll(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func main() {
+
+	client, err := unversioned.NewInCluster()
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	ipt := iptables.New(exec.New(), dbus.New(), iptables.ProtocolIpv4)
+	ips := ipset.New()
+
+	handleError(resetIPTables(ipt))
+	handleError(resetIPSets(ips))
+
+	npc := weavenpc.New(ipt, ips)
+
+	nsController := makeController(client, "namespaces", &api.Namespace{},
+		framework.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				handleError(npc.AddNamespace(obj.(*api.Namespace)))
+			},
+			DeleteFunc: func(obj interface{}) {
+				handleError(npc.DeleteNamespace(obj.(*api.Namespace)))
+			},
+			UpdateFunc: func(old, new interface{}) {
+				handleError(npc.UpdateNamespace(old.(*api.Namespace), new.(*api.Namespace)))
+			}})
+
+	podController := makeController(client, "pods", &api.Pod{},
+		framework.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				handleError(npc.AddPod(obj.(*api.Pod)))
+			},
+			DeleteFunc: func(obj interface{}) {
+				handleError(npc.DeletePod(obj.(*api.Pod)))
+			},
+			UpdateFunc: func(old, new interface{}) {
+				handleError(npc.UpdatePod(old.(*api.Pod), new.(*api.Pod)))
+			}})
+
+	npController := makeController(client.ExtensionsClient, "networkpolicies", &extensions.NetworkPolicy{},
+		framework.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				handleError(npc.AddNetworkPolicy(obj.(*extensions.NetworkPolicy)))
+			},
+			DeleteFunc: func(obj interface{}) {
+				handleError(npc.DeleteNetworkPolicy(obj.(*extensions.NetworkPolicy)))
+			},
+			UpdateFunc: func(old, new interface{}) {
+				handleError(npc.UpdateNetworkPolicy(old.(*extensions.NetworkPolicy), new.(*extensions.NetworkPolicy)))
+			}})
+
+	go nsController.Run(wait.NeverStop)
```

Comment:
There's a mutex in the `NetworkPolicyController` that serialises event processing.